text/0000-app-namespacing-semantic-convention.md

```diff
@@ -0,0 +1,61 @@
+# Application namespacing semantic convention
+
+Adds a semantic convention for manual instrumentation that defines how to namespace custom attributes, starting with the `app` prefix.
+
+## Motivation
+
+Today, custom attributes from an app created through manual instrumentation have all kinds of names that are specific to the domain of that app. There are no conventions for how to organize this data, though.
+
+Data organization is critical to being able to effectively analyze observability data and avoid collisions in names. Without a well-defined convention for developers to follow, their own means of organization can differ from developer to developer, leading to a need to do things like coalesce field names using `.`/`-`/`_`/etc, not to mention having a real possibility to have name collisions. to separate clauses in an attribute name. Instead, a well-documented convention for how to organize data and in what format removes a "need to think" like code formatting conventions can do.
+
+## Explanation
+
+Attributes created through manual instrumentation should use namespacing for organizational purposes, using `app.` as a prefix for the name. Sub-namespaces should use a period (`.`) to separate sub-namespaces.
+
+Use as many layers of subnamespaces as needed to represent your app's hierarchy. For example, an eCommerce app may define attributes like this:
+
+* `app.order.id`
+* `app.cart.items.count`
+* `app.shipping.cost.total`
```

Comment:
Not sure if I am missing something but it is not clear to me why these attributes are not prefixed with `com.<company>.`. If a customer chooses to do manual instrumentation and introduce attributes that the telemetry backends do not have a _standard_ interpretation for, then they fit the definition of custom attributes.  That is, they need to tell the telemetry backend how to interpret these attributes, say through configuration.

In this example of eCommerce app, since it is common enough I believe we can add them to the semantic conventions in the spec and still use them from manual instrumentation.

Reply:
@Oberon00 made a similar comment [above](https://github.com/open-telemetry/oteps/pull/211#discussion_r948154299). Copying from the spec:

> * The name is specific to your company and may be possibly used outside the company as well. To avoid clashes with names introduced by other companies (in a distributed system that uses applications from multiple vendors) it is recommended to prefix the new name by your company's reverse domain name, e.g. com.acme.shopname.
> * The name is specific to your application that will be used internally only. If you already have an internal company process that helps you to ensure no name clashes happen then feel free to follow it. Otherwise it is recommended to prefix the attribute name by your application name, provided that the application name is reasonably unique within your organization (e.g. myuniquemapapp.longitude is likely fine). Make sure the application name does not clash with an existing semantic convention namespace.
> * The name may be generally applicable to applications in the industry. In that case consider submitting a proposal to this specification to add a new name to the semantic conventions, and if necessary also to add a new namespace.

All of those are suggestions (at least I don't read any enforcing language there). I read some of the OTEP and some of the comments here, that it would be good to **enforce** a namespace for applications, e.g. to avoid a clash not only with existing but also with future namespaces.

I think that the `com.companyname.whatever` is already a good approach, but I see one issue with that: imagine you have acquired two pieces of off-the-shelf software, both have opentelemetry and send traces&metrics. Now both send you some information like the `order.id` with their spans/traces. Now the data emitted from software 1 is `com.vendor1.order.id` and the data emitted from software 2 is `com.vendor2.order.id`, although both might be the same. So what you might want to have is a `heresmymagixprefix.order.id` in both cases.

Reply:
@svrnm what do you have in mind on how the backend should interpret the custom attributes. My understanding is that the customers should convey that to the backends. So in this case, they would have to convey `com.vendor1.order.id` and `com.vendor1.order.id` as being for the same purpose.